File: racingmanager-backend/src/main/kotlin/io/github/raginlundf/racingmanager/api/spectator/models/SpectatorModels.kt
package io.github.raginlundf.racingmanager.api.spectator.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class SpectatorSnapshotResponseModel(
    val event: SpectatorEventModel,
    val currentHeat: SpectatorHeatModel? = null,
    val upcomingHeats: List<SpectatorHeatModel> = emptyList(),
    val qualificationRankings: List<SpectatorRankingEntryModel> = emptyList(),
    val qualificationStatus: String? = null,
    val knockout: SpectatorKnockoutStateModel? = null,
)

@Serializable
data class SpectatorEventModel(
    val id: String,
    val name: String,
    val description: String? = null,
    val status: String,
    val laneType: String,
    val measurementType: String,
)

@Serializable
data class SpectatorHeatModel(
    val id: String,
    val heatNumber: Int,
    val round: Int,
    val status: String,
    val lanes: List<SpectatorLaneModel>,
    val hasResult: Boolean = false,
)

@Serializable
data class SpectatorLaneModel(
    val lane: Int,
    val participantId: String,
    val participantStartNumber: Int,
    val participantFirstName: String,
    val participantLastName: String,
    val durationNanos: Long? = null,
    val outcome: String? = null,
)

@Serializable
data class SpectatorRankingEntryModel(
    val participantId: String,
    val startNumber: Int,
    val firstName: String,
    val lastName: String,
    val club: String? = null,
    val bestTimeNanos: Long? = null,
    val totalTimeNanos: Long? = null,
    val completedRuns: Int = 0,
    val dnfCount: Int = 0,
    val rank: Int = 0,
)

@Serializable
data class SpectatorKnockoutStateModel(
    val status: String,
    val pairingMode: String,
    val rounds: List<SpectatorKnockoutRoundModel> = emptyList(),
)

@Serializable
data class SpectatorKnockoutRoundModel(
    val roundNumber: Int,
    val matches: List<SpectatorKnockoutMatchModel> = emptyList(),
)

@Serializable
data class SpectatorKnockoutMatchModel(
    val id: String,
    val roundNumber: Int,
    val matchNumber: Int,
    val participant1Id: String? = null,
    val participant2Id: String? = null,
    val winnerId: String? = null,
    val status: String,
    val isBye: Boolean = false,
)

@Serializable
data class SpectatorTokenResponseModel(
    val exchangeCode: String,
    val expiresIn: Long,
)

@Serializable
data class SpectatorExchangeRequestModel(
    val code: String,
)

@Serializable
data class SpectatorExchangeResponseModel(
    val accessToken: String,
    val expiresIn: Long,
    val eventId: String,
)

File: racingmanager-webapp/src/app/libs/clients/spectator/spectator.models.ts
export interface SpectatorTokenResponse {
  exchangeCode: string;
  expiresIn: number;
}

export interface SpectatorExchangeResponse {
  accessToken: string;
  expiresIn: number;
  eventId: string;
}

export interface SpectatorEventModel {
  id: string;
  name: string;
  description?: string;
  status: string;
  laneType: string;
  measurementType: string;
}

export interface SpectatorLaneModel {
  lane: number;
  participantId: string;
  participantStartNumber: number;
  participantFirstName: string;
  participantLastName: string;
  durationNanos?: number;
  outcome?: string;
}

export interface SpectatorHeatModel {
  id: string;
  heatNumber: number;
  round: number;
  status: string;
  lanes: SpectatorLaneModel[];
  hasResult: boolean;
}

export interface SpectatorRankingEntryModel {
  participantId: string;
  startNumber: number;
  firstName: string;
  lastName: string;
  club?: string;
  bestTimeNanos?: number;
  totalTimeNanos?: number;
  completedRuns: number;
  dnfCount: number;
  rank: number;
}

export interface SpectatorKnockoutMatchModel {
  id: string;
  roundNumber: number;
  matchNumber: number;
  participant1Id?: string;
  participant2Id?: string;
  winnerId?: string;
  status: string;
  isBye: boolean;
}

export interface SpectatorKnockoutRoundModel {
  roundNumber: number;
  matches: SpectatorKnockoutMatchModel[];
}

export interface SpectatorKnockoutStateModel {
  status: string;
  pairingMode: string;
  rounds: SpectatorKnockoutRoundModel[];
}

export interface SpectatorParticipantStandingModel {
  participantId: string;
  startNumber: number;
  firstName: string;
  lastName: string;
  bestQualificationTimeNanos?: number;
  bestKnockoutTimeNanos?: number;
  state: string;
  place: number;
  racing: boolean;
}

export interface SpectatorSnapshotResponse {
  event: SpectatorEventModel;
  currentHeat?: SpectatorHeatModel;
  upcomingHeats: SpectatorHeatModel[];
  qualificationRankings: SpectatorRankingEntryModel[];
  qualificationStatus?: string;
  knockout?: SpectatorKnockoutStateModel;
  knockoutStandings?: SpectatorParticipantStandingModel[];
}
